pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -36,260 +36,107 @@ var _ apis.Validatable = (*Pipeline)(nil)
+func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
 	if equality.Semantic.DeepEqual(ps, &PipelineSpec{}) {
-		return apis.ErrGeneric("expected at least one, got none", "spec.description", "spec.params", "spec.resources", "spec.tasks", "spec.workspaces")
+		errs = errs.Also(apis.ErrGeneric("expected at least one, got none", "description", "params", "resources", "tasks", "workspaces"))
 	}
-
 	// PipelineTask must have a valid unique label and at least one of taskRef or taskSpec should be specified
-	if err := validatePipelineTasks(ctx, ps.Tasks, ps.Finally); err != nil {
-		return err
-	}
-
+	errs = errs.Also(validatePipelineTasks(ctx, ps.Tasks, ps.Finally))
 	// All declared resources should be used, and the Pipeline shouldn't try to use any resources
 	// that aren't declared
-	if err := validateDeclaredResources(ps.Resources, ps.Tasks, ps.Finally); err != nil {
-		return apis.ErrInvalidValue(err.Error(), "spec.resources")
-	}
-
+	errs = errs.Also(validateDeclaredResources(ps.Resources, ps.Tasks, ps.Finally))
 	// The from values should make sense
-	if err := validateFrom(ps.Tasks); err != nil {
-		return err
-	}
-
+	errs = errs.Also(validateFrom(ps.Tasks))
 	// Validate the pipeline task graph
-	if err := validateGraph(ps.Tasks); err != nil {
-		return apis.ErrInvalidValue(err.Error(), "spec.tasks")
-	}
-
-	if err := validateParamResults(ps.Tasks); err != nil {
-		return apis.ErrInvalidValue(err.Error(), "spec.tasks.params.value")
-	}
-
+	errs = errs.Also(validateGraph(ps.Tasks))
+	errs = errs.Also(validateParamResults(ps.Tasks))
 	// The parameter variables should be valid
-	if err := validatePipelineParameterVariables(ps.Tasks, ps.Params); err != nil {
-		return err
-	}
-
-	if err := validatePipelineParameterVariables(ps.Finally, ps.Params); err != nil {
-		return err
-	}
-
-	if err := validatePipelineContextVariables(ps.Tasks); err != nil {
-		return err
-	}
-
+	errs = errs.Also(validatePipelineParameterVariables(ps.Tasks, ps.Params).ViaField("tasks"))
+	errs = errs.Also(validatePipelineParameterVariables(ps.Finally, ps.Params).ViaField("finally"))
+	errs = errs.Also(validatePipelineContextVariables(ps.Tasks))
 	// Validate the pipeline's workspaces.
-	if err := validatePipelineWorkspaces(ps.Workspaces, ps.Tasks, ps.Finally); err != nil {
-		return err
-	}
-
+	errs = errs.Also(validatePipelineWorkspaces(ps.Workspaces, ps.Tasks, ps.Finally))
 	// Validate the pipeline's results
-	if err := validatePipelineResults(ps.Results); err != nil {
-		return apis.ErrInvalidValue(err.Error(), "spec.tasks.params.value")
-	}
-
-	if err := validateTasksAndFinallySection(ps); err != nil {
-		return err
-	}
-
-	if err := validateFinalTasks(ps.Finally); err != nil {
-		return err
-	}
-
-	if err := validateWhenExpressions(ps.Tasks); err != nil {
-		return err
-	}
-
-	return nil
+	errs = errs.Also(validatePipelineResults(ps.Results))
+	errs = errs.Also(validateTasksAndFinallySection(ps))
```

Comment:
This is a bit duplicated (in some weird way) with the task validation above. The whole `task` and `finally` validation code could be enhanced in follow-ups